File: CPP_Files/Euler.cpp
#include <vector>
#include <functional>

class EulerSolver {
private:
    std::vector<double> solutions_;
    size_t num_steps_;
    size_t num_vars_;

    size_t idx(size_t step, size_t var) const {
        return step * num_vars_ + var;
    }

public:
    EulerSolver(size_t num_steps, size_t num_vars)
        : num_steps_(num_steps), num_vars_(num_vars) {
            solutions_.resize(num_steps * num_vars, 0.0);
        }

    void solve(const std::vector<double>& initial_conditions,
               double dt,
               std::function<std::vector<double>(double, const std::vector<double>&)> derivative) {
        
        // Set initial conditions
        for (size_t i = 0; i < num_vars_; ++i) {
            at(0, i) = initial_conditions[i];
        }

        // Euler integration
        for (size_t t = 0; t < num_steps_ - 1; ++t) {
            double current_time = t * dt;  
            auto deriv = derivative(current_time, get_state(t)); 

            for (size_t i = 0; i < num_vars_; ++i) {
                at(t + 1, i) = at(t, i) + dt * deriv[i]; 
            }
        }
    }

    double& at(size_t step, size_t var) { 
        return solutions_[idx(step, var)];
    }

    double at(size_t step, size_t var) const {  
        return solutions_[idx(step, var)];
    }

    std::vector<double> get_state(size_t step) const {
        std::vector<double> state(num_vars_);
        for (size_t i = 0; i < num_vars_; ++i) {
            state[i] = solutions_[idx(step, i)];
        }
        return state;
    }
};

File: Computation/Numerical Methods/Euler.py
import numpy as np
from HH_ODE import hh_ode
import matplotlib.pyplot as plt

def euler_method(f, t_span, y0, n_steps, *args):
    """
    Function Description:
        - Solves a system of ordinary differential equations using Euler's method
        - Calculates the tangent slope at each time step to approximate the solution
    
    Parameters:
        - f (callable): Function that computes the derivatives of the system
            - Signature: f(t, y, *args) -> array_like
        - t_span (tuple): Time range for integration as (start_time, end_time)
        - y0 (array_like): Initial conditions vector
        - n_steps (int): Number of time steps to divide the integration interval
        - *args: Additional arguments to pass to the ODE function f
    
    Returns:
        - t (numpy.ndarray): Array of time points from t_span[0] to t_span[1]
        - y (numpy.ndarray): Solution matrix containing approximated state variables
    """
    
    # Create time array
    t = np.linspace(t_span[0], t_span[1], n_steps + 1)
    dt = t[1] - t[0]
    
    # Initialize solution array
    y = np.full((len(t), len(y0)), np.nan)
    y[0, :] = y0

    # Iteratively approximate the solution using Euler's method
    for i in range(len(t) - 1):
        # Calculate derivative at current point and update
        derivative = f(t[i], y[i, :], *args)
        y[i + 1, :] = y[i, :] + dt * derivative

    return t, y


if __name__ == "__main__":

    t_span = (0, 50)  # ms
    y0 = np.array([-65, 0.05, 0.6, 0.32])  # [V0, m0, h0, n0]
    n_steps = 1000
    
    def I_ext(t):
        return 10.0 if 10 <= t <= 40 else 0.0
    
    # Solve using Euler method
    t, solution = euler_method(hh_ode, t_span, y0, n_steps, I_ext)
    
    # Plot results
    plt.figure(figsize=(12, 8))
    
    plt.subplot(2, 2, 1)
    plt.plot(t, solution[:, 0])
    plt.title('Membrane Potential V(t)')
    plt.xlabel('Time (ms)')
    plt.ylabel('V (mV)')
    
    plt.subplot(2, 2, 2)
    plt.plot(t, solution[:, 1], label='m')
    plt.plot(t, solution[:, 2], label='h')
    plt.plot(t, solution[:, 3], label='n')
    plt.title('Gating Variables')
    plt.xlabel('Time (ms)')
    plt.ylabel('Gating variable')
    plt.legend()
    
    plt.tight_layout()
    plt.show()